File: backend/app/models/tag.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..core.database import Base
from .document import document_tags


class Tag(Base):
    __tablename__ = "tags"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    color: Mapped[str] = mapped_column(String(7), default="#3B82F6")  # hex color
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)

    documents = relationship(
        "Document",
        secondary=document_tags,
        back_populates="tags",
    )

File: frontend/src/types/tag.ts
export type Tag = {
  id: number;
  name: string;
  color: string;
  document_count: number;
  created_at: string;
};

export type TagCreateInput = {
  name: string;
  color: string;
};

export function tagToSelectOption(tag: Tag): { label: string; value: number } {
  return { label: tag.name, value: tag.id };
}

